From 5f91f81f196cb9d0dbb460f0632b486f1b1fdc1a Mon Sep 17 00:00:00 2001
From: Biswarup Pal <biswarupp@google.com>
Date: Wed, 29 Apr 2026 12:04:35 -0700
Subject: [PATCH 29/49] Restrict VirtualDeviceImpl methods to device owner.

Add checkCallerIsDeviceOwner() to setListeners, close, getVirtualCameraId, and playSoundEffect to ensure only the owning application can call these methods.

Test: presubmit
Bug: 495646742
Flag: EXEMPT BUGFIX
(cherry picked from commit fe4e7456e4353ce846e3e72169f86ba04020f286)
Cherrypick-From: https://googleplex-android-review.googlesource.com/q/commit:a101654da176c6a855f134eedfcf893d698da329
Merged-In: I68d8cdb0de58ffc2c4109264e0a49b078ae0c17e
Change-Id: I68d8cdb0de58ffc2c4109264e0a49b078ae0c17e
---
 .../android/server/companion/virtual/VirtualDeviceImpl.java   | 4 ++++
 .../server/companion/virtual/VirtualDeviceShellCommand.java   | 2 +-
 2 files changed, 5 insertions(+), 1 deletion(-)

diff --git a/services/companion/java/com/android/server/companion/virtual/VirtualDeviceImpl.java b/services/companion/java/com/android/server/companion/virtual/VirtualDeviceImpl.java
index 850e8806e698..35217902eaaf 100644
--- a/services/companion/java/com/android/server/companion/virtual/VirtualDeviceImpl.java
+++ b/services/companion/java/com/android/server/companion/virtual/VirtualDeviceImpl.java
@@ -730,6 +730,7 @@ final class VirtualDeviceImpl extends IVirtualDevice.Stub implements IBinder.Dea
     @Override // Binder call
     public void setListeners(@NonNull IVirtualDeviceActivityListener activityListener,
             @NonNull IVirtualDeviceSoundEffectListener soundEffectListener) {
+        checkCallerIsDeviceOwner();
         mActivityListener = Objects.requireNonNull(activityListener);
         mSoundEffectListener = Objects.requireNonNull(soundEffectListener);
     }
@@ -949,6 +950,7 @@ final class VirtualDeviceImpl extends IVirtualDevice.Stub implements IBinder.Dea
 
     @Override // Binder call
     public void close() {
+        checkCallerIsDeviceOwner();
         // Remove about-to-be-closed virtual device from the service before butchering it.
         if (!mService.removeVirtualDevice(mDeviceId)) {
             // Device is already closed.
@@ -1394,6 +1396,7 @@ final class VirtualDeviceImpl extends IVirtualDevice.Stub implements IBinder.Dea
     @Override // Binder call
     public String getVirtualCameraId(@NonNull VirtualCameraConfig cameraConfig)
             throws RemoteException {
+        checkCallerIsDeviceOwner();
         Objects.requireNonNull(cameraConfig);
         if (mVirtualCameraController == null) {
             throw new UnsupportedOperationException("Virtual camera controller is not available");
@@ -1967,6 +1970,7 @@ final class VirtualDeviceImpl extends IVirtualDevice.Stub implements IBinder.Dea
     }
 
     void playSoundEffect(int effectType) {
+        checkCallerIsDeviceOwner();
         if (mSoundEffectListener == null) {
             return;
         }
diff --git a/services/companion/java/com/android/server/companion/virtual/VirtualDeviceShellCommand.java b/services/companion/java/com/android/server/companion/virtual/VirtualDeviceShellCommand.java
index 410e80a77fe2..4f170f51f27d 100644
--- a/services/companion/java/com/android/server/companion/virtual/VirtualDeviceShellCommand.java
+++ b/services/companion/java/com/android/server/companion/virtual/VirtualDeviceShellCommand.java
@@ -166,7 +166,7 @@ class VirtualDeviceShellCommand extends ShellCommand {
         if (virtualDevice == null) {
             return 1;
         }
-        virtualDevice.close();
+        Binder.withCleanCallingIdentity(() -> virtualDevice.close());
         return 0;
     }
 
-- 
2.55.0

